File: 0528-random-pick-with-weight/0528-random-pick-with-weight.cpp
class Solution {
public:
    vector<int> prefixSum;
    Solution(vector<int>& w) {
        int total = 0;
        
        for(auto& num : w)
        {
            total+=num;
            prefixSum.push_back(total);
        }
    }
    
    int pickIndex() {
        float r = (float) rand() / RAND_MAX;
        float target = prefixSum.back() * r;
        return lower_bound(prefixSum.begin(), prefixSum.end(), target) - prefixSum.begin();
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(w);
 * int param_1 = obj->pickIndex();
 */

File: 0528-random-pick-with-weight/0528-random-pick-with-weight.py
class Solution:

    def __init__(self, w: List[int]):
        self.prefix = []
        total = 0
        for weight in w:
            total += weight
            self.prefix.append(total)
            
        # print((self.prefix))

    def pickIndex(self) -> int:
        r = random.random() * self.prefix[-1]
        
        for i, num in enumerate(self.prefix):
            if r <= num:
                return i


# Your Solution object will be instantiated and called as such:
# obj = Solution(w)
# param_1 = obj.pickIndex()